### DISCUSSION TITLE
How do I detect in a Theia extension whether a document has been modified and saved?

### DISCUSSION THREAD

**Discussion by tsiyona** [eclipse-theia/theia#9041] [Q&A] [answered]
We wrote a Theia extension, and we would like to receive an event every time a document (java source file, for example) is modified and **saved**. Incremental changes before the actual Save are irrelevant for us.
How can we be notified about a code change?

**Comment by vince-fugnitto** [comment, marked as answer]
@tsiyona you can listen to the `onDidSaveTextDocument` event (documents are represented by underlying `models`):

https://github.com/eclipse-theia/theia/blob/27e6ea70789cc34c06d1a1e1d52d3aee2afb902c/packages/monaco/src/browser/monaco-workspace.ts#L85

From the event you'll have access to the [`MonacoEditorModel`](https://eclipse-theia.github.io/theia/docs/next/classes/monaco.monacoeditormodel.html) from which you can have special handling for `java` files as example if that's your use-case.

**Reply by tsiyona** [reply]
Thanks for your help! I implemented what you suggested, and indeed, I receive an event with MonacoEditorModel upon Save. However, when I write in the document but don't yet press Save (I just pause for a second), I also get the event. 
Is this normal behavior?
Maybe there is an auto-save in the background. I will check.

**Reply by vince-fugnitto** [reply]
@tsiyona can you confirm the value of the `editor.autoSave` preference? (confirm the preference at the `user` and `workspace` scope). Verifying the code and when the event is fired it looks like it only occurs during actual saves, not when the document is modified.

**Reply by tsiyona** [reply]
Yes @vince-fugnitto , _editor.autoSave_ was 'on' with an _autoSaveDelay_ of 500 ms. That's why it was firing the event. I turned autoSave 'off' and now only upon Save does the event fire. It's working well. Thank you!
I did not understand how I can set the autoSave preference at the _user_ vs. _workspace_ scope though.

**Reply by vince-fugnitto** [reply]
@tsiyona no problem I'm happy to help! 


For your own understanding (unrelated), the framework supports the following preference scopes (similarly to vscode):
- `user`: the `~/.theia/settings.json` under the user's home 
- `workspace`: the `.theia/settings.json` under the workspace folder (we also read `.vscode/settings.json`)
- `folder`: in a multiple-root workspace, similar to `workspace`.

The behavior is the same as in vscode:
- https://code.visualstudio.com/docs/getstarted/settings

**Reply by tsiyona** [reply]
Thank you. Just one more question.
I see that changes I make in the UI via _File -> Settings -> Open Preferences_ are saved in the _user_ preferences.
How do I change the _workspace_ settings? Do I edit the settings file manually, or can it also be done via the UI?

**Reply by vince-fugnitto** [reply]
@tsiyona when opening the _Preferences UI_, the default scope is for _User_. Given that a user has the workspace opened, you can select the _Workspace_ tab to update preferences for the workspace. This will create the `.theia/settings.json` file if necessary, and add your updated preference. You can also manually edit the `settings.json` file and it will give you schema, and autocompletion support.

<img width="1355" alt="Screen Shot 2021-02-09 at 6 52 20 PM" src="https://user-images.githubusercontent.com/40359487/107444598-d86f1100-6b08-11eb-99fd-093ca1b0508e.png">

**Reply by tsiyona** [reply]
Going back to onDidSaveTextDocument event. How can I know what line numbers were changed?

All in all, what I need to know about a modified file is:

1. File new content.
2. File full path.
3. File language (java, for now).
3. Fully qualified name of the class in the changed lines (for *.java files).

What I already have:
- The file content and full path indeed appear in MonacoEditorModel,
- The language I can detect by the file extension, though I see that MonacoEditorModel does include a __languageId_ member.

What I am still missing:
- The line numbers of the modified lines in the file.
If I have the file path and the line numbers, I can find the fully qualified name.

**Reply by vince-fugnitto** [reply]
@tsiyona you'll need to take a look at the API and events to see what might give you the information you require. 

- https://eclipse-theia.github.io/theia/docs/next/classes/monaco.monacoworkspace.html

In the case of getting the actual edit content, the `save` event is likely too late and does not have the information you need for your use-case. You can take a look at the `onDidChangeTextDocument` event which contains the `contentChanges` from which you can determine what line was updated (`start` and `end`).

Example:

```ts
this.monacoWorkspace.onDidChangeTextDocument(editor => {
    // For this example we only care about dirty (unsaved) changes in the document.
    if (editor.model.dirty) {
        const changes = editor.contentChanges;
        changes.forEach(change => {
            if ('range' in change) {
                const range = change.range;
                console.log(`line number: ${range.start.line + 1}`);
            }
        });
    }
});
```

**Reply by tsiyona** [reply]
Thank you @vince-fugnitto , in fact, I already implemented listening on _onWillSaveTextDocument_, however I suspect that I am getting line numbers also for edit changes that were reverted (Ctrl-Z). Apparently this is happening because I did not check the dirty flag.

```
constructor() {
	this.monacoWorkspace.onWillSaveTextDocument(this.onWillSaveTextDocument);
}

onWillSaveTextDocument = async (willSaveEvent: MonacoTextDocumentWillSaveEvent) => {
	const path = willSaveEvent.textDocument.uri;
	const contentChanges = new Set(willSaveEvent.textDocument.contentChanges.map(v => {return v.range.start.line}));
	const content = willSaveEvent.textDocument.getText();
	const fileExt = path.split('.').pop();

	// use extracted info...
}

```

Comparing it to your code snippet, I
- should have checked _willSaveEvent._textDocument.dirty__ flag.
- should have also checked _if ('range' in change)_
- should have added 1 to the line number.

Anyway, I will try using _onDidChangeTextDocument_.

Thank you!!
